// @smithery/sdk/dist/shared/config.js
import _ from "lodash";
import { err, ok } from "okay-error";
import { zodToJsonSchema } from "zod-to-json-schema";
function isPlainObject(value) {
    return value !== null && typeof value === "object" && !Array.isArray(value);
}
export function appendConfigAsDotParams(url, config) {
    function add(pathParts, value) {
        if (Array.isArray(value)) {
            for (let index = 0; index < value.length; index++) {
                add([...pathParts, String(index)], value[index]);
            }
            return;
        }
        if (isPlainObject(value)) {
            for (const [key, nested] of Object.entries(value)) {
                add([...pathParts, key], nested);
            }
            return;
        }
        const key = pathParts.join(".");
        let stringValue;
        switch (typeof value) {
            case "string":
                stringValue = value;
                break;
            case "number":
            case "boolean":
                stringValue = String(value);
                break;
            default:
                stringValue = JSON.stringify(value);
        }
        url.searchParams.set(key, stringValue);
    }
    if (isPlainObject(config)) {
        for (const [key, value] of Object.entries(config)) {
            add([key], value);
        }
    }
}
/**
 * Creates a URL to connect to the Smithery MCP server.
 * @param baseUrl The base URL of the Smithery server
 * @param options Optional configuration object
 * @returns A URL with config encoded using dot-notation query params (e.g. model.name=gpt-4&debug=true)
 */
export function createSmitheryUrl(baseUrl, options) {
    const url = new URL(`${baseUrl}/mcp`);
    if (options?.config) {
        appendConfigAsDotParams(url, options.config);
    }
    if (options?.apiKey) {
        url.searchParams.set("api_key", options.apiKey);
    }
    if (options?.profile) {
        url.searchParams.set("profile", options.profile);
    }
    return url;
}
/**
 * Parses and validates config from an Express request with optional Zod schema validation
 * Supports dot-notation config parameters (e.g., foo=bar, a.b=c)
 * @param req The express request
 * @param schema Optional Zod schema for validation
 * @returns Result with either parsed data or error response
 */
export function parseAndValidateConfig(req, schema) {
    const config = parseConfigFromQuery(Object.entries(req.query));
    // Validate config against schema if provided
    if (schema) {
        const result = schema.safeParse(config);
        if (!result.success) {
            const jsonSchema = zodToJsonSchema(schema);
            const errors = result.error.issues.map(issue => {
                // Safely traverse the config object to get the received value
                let received = config;
                for (const key of issue.path) {
                    if (received && typeof received === "object" && key in received) {
                        received = received[key];
                    }
                    else {
                        received = undefined;
                        break;
                    }
                }
                return {
                    param: issue.path.join(".") || "root",
                    pointer: `/${issue.path.join("/")}`,
                    reason: issue.message,
                    received,
                };
            });
            return err({
                title: "Invalid configuration parameters",
                status: 422,
                detail: "One or more config parameters are invalid.",
                instance: req.originalUrl,
                configSchema: jsonSchema,
                errors,
            });
        }
        return ok(result.data);
    }
    return ok(config);
}
// Process dot-notation config parameters from query parameters (foo=bar, a.b=c)
// This allows URL params like ?server.host=localhost&server.port=8080&debug=true
export function parseConfigFromQuery(query) {
    const config = {};
    for (const [key, value] of query) {
        // Skip reserved parameters
        if (key === "api_key" || key === "profile")
            continue;
        const pathParts = key.split(".");
        // Handle array values from Express query parsing
        const rawValue = Array.isArray(value) ? value[0] : value;
        if (typeof rawValue !== "string")
            continue;
        // Try to parse value as JSON (for booleans, numbers, objects)
        let parsedValue = rawValue;
        try {
            parsedValue = JSON.parse(rawValue);
        }
        catch {
            // If parsing fails, use the raw string value
        }
        // Use lodash's set method to handle nested paths
        _.set(config, pathParts, parsedValue);
    }
    return config;
}
