// @smithery/registry/dist/esm/models/errors/servererror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { SmitheryRegistryError } from "./smitheryregistryerror.js";
/**
 * An unexpected error occurred
 */
export class ServerError extends SmitheryRegistryError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        if (err.error != null)
            this.error = err.error;
        this.name = "ServerError";
    }
}
/** @internal */
export const ServerError$inboundSchema = z.object({
    error: z.string().optional(),
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new ServerError(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
/** @internal */
export const ServerError$outboundSchema = z.instanceof(ServerError)
    .transform(v => v.data$)
    .pipe(z.object({
    error: z.string().optional(),
}));
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var ServerError$;
(function (ServerError$) {
    /** @deprecated use `ServerError$inboundSchema` instead. */
    ServerError$.inboundSchema = ServerError$inboundSchema;
    /** @deprecated use `ServerError$outboundSchema` instead. */
    ServerError$.outboundSchema = ServerError$outboundSchema;
})(ServerError$ || (ServerError$ = {}));
