// @smithery/registry/dist/esm/models/errors/unauthorizederror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { SmitheryRegistryError } from "./smitheryregistryerror.js";
/**
 * Authentication information is missing or invalid
 */
export class UnauthorizedError extends SmitheryRegistryError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        if (err.error != null)
            this.error = err.error;
        this.name = "UnauthorizedError";
    }
}
/** @internal */
export const UnauthorizedError$inboundSchema = z.object({
    error: z.string().optional(),
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new UnauthorizedError(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
/** @internal */
export const UnauthorizedError$outboundSchema = z.instanceof(UnauthorizedError)
    .transform(v => v.data$)
    .pipe(z.object({
    error: z.string().optional(),
}));
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var UnauthorizedError$;
(function (UnauthorizedError$) {
    /** @deprecated use `UnauthorizedError$inboundSchema` instead. */
    UnauthorizedError$.inboundSchema = UnauthorizedError$inboundSchema;
    /** @deprecated use `UnauthorizedError$outboundSchema` instead. */
    UnauthorizedError$.outboundSchema = UnauthorizedError$outboundSchema;
})(UnauthorizedError$ || (UnauthorizedError$ = {}));
